### PR TITLE
Add version class and update FileType classes

### PR DESCRIPTION
**Pull request by coppolat1**
Add class to encapsulate version number instead of using an array of strings, move lots of repeated logic here. Refactor FileType children to be more concise. Update content replacement code. Formatiting, variable name updates.

### CODE REVIEW

File: filetypes.py [hunk 1]
```diff
@@ -3,40 +3,49 @@
 from exceptions import DoxyException, PreProcessorException
 
 
-class Filetype():
-
-    version = ["0", "0", "0"]  # [major, minor, patch]
-    # Namespace(config_file='./tests/.bump.cfg', version_file=None, part='minor', dont_reset=False, component='naibrd')
-    args = None
-    target_file = None  # init as None
+class SemanticVersionNumber(object):
```
Comment (travisrecupero) [inline filetypes.py:6]: Much better than version array!

File: filetypes.py [hunk 2]
```diff
@@ -62,85 +71,101 @@ class PreProcessor(Filetype):
...
-        # get major
-        matchMaj = re.search(self.r_major, content, re.X)
-        # get minor
-        matchMin = re.search(self.r_minor, content, re.X)
-        # get patch
-        matchPat = re.search(self.r_patch, content, re.X)
+        # Get match objects and save them
+        self._match_obj_major = re.search(self.r_major, content, re.X)
+        self._match_obj_minor = re.search(self.r_minor, content, re.X)
+        self._match_obj_patch = re.search(self.r_patch, content, re.X)
```
Comment (travisrecupero) [inline filetypes.py:84]: Is  `self.` in `self._match_obj_major`, `self._match_obj_minor`, `self._match_obj_patch`  needed here?

File: filetypes.py [hunk 3]
```diff
@@ -62,85 +71,101 @@ class PreProcessor(Filetype):
...
+                match_obj.group('varName') +
+                match_obj.group('nSpaces') + '(' +
+                str(new_version) +
+                match_obj.group('unsigned') + ')')
 
 
 class Doxy(Filetype):
     #   Define regex patterns
-    r_pattern = r"PROJECT_NUMBER\s*=\s*(?P<major>\d+)\.(?P<minor>\d+)?\.(?P<patch>\*|\d+)"
+    r_pattern = r"PROJECT_NUMBER\s?=\s?(?P<major>\d+)\.(?P<minor>\d+)?\.(?P<patch>\*|\d+)"
```
Comment (travisrecupero) [inline filetypes.py]: This does not work if Doxy uses tabs between tokens

**Review by travisrecupero — CHANGES_REQUESTED** [review]
## Things to address
- regex using `*` instead of `?`.  Discuss before merging.

## Comments
Other than that it looks good to me! Writing to files now seems less prone to error since we are no longer reading line by line, but rather using a custom replace function inside `re.sub(..)`.

File: filetypes.py (hunk 3 above)
Comment (coppolat1) [reply filetypes.py]: By tabs between tokens, do you mean `PROJECT_NUMBER\t=\tmajor.minor.patch`?

Comment (travisrecupero) [reply filetypes.py]: Yes so,
PROJECT_NUMBER &emsp; = &emsp; 1.0.0     <- doesnt work
vs
PROJECT_NUMBER = 1.0.0 <- works
vs
PROJECT_NUMBER=1.0.0 <- works

Comment (coppolat1) [reply filetypes.py]: Okay, I figured it out. The first one you mentioned doesn't just have tabs in-between the equal sign, it has tabs surrounded by spaces:
`PROJECT_NUMBER \t = \t 1.0.0`

The question mark in `\s?` means 'zero or one' white-space.. but we have two in that particular case, so no match. So, I'm not sure if having tabs+spaces in-between the equal sign is something that we need to expect, but to make it work for all cases I'll change the `?` to a `*`.

File: filetypes.py (hunk 2 above)
Comment (coppolat1) [reply filetypes.py:84]: Yes, unfortunately we have to use `self` when referencing a class attribute from inside a classes method.

**Comment by coppolat1** [comment]
Thanks for all the comments, Travis. I fixed the regex issue and will merge this branch in shortly.

**Pull request merged**